core/ir/synthetic/encoding_common.h

```diff
@@ -0,0 +1,147 @@
+ * destination register operands, register operands and their size (which is the same for
+ * all operands), instruction length, and ISA mode (which is DR_ISA_REGDEPS).
+ * Querying additional information (e.g., the instruction opcode) will return the default
+ * value generated by instr_create() (i.e., zero).
+ */
+
+/* XXX i#1684: Note that this is part of a larger issue, where lack of cross-arch support
+ * in the same build of DynamoRIO is limiting us.
+ */
+
+/* Here we describe the encoding scheme for the Synthetic ISA that is enforced in decode.c
+ * and encode.c.
+ *
+ * Encoded instructions are 4 byte aligned.
+ *
+ * All instruction encodings begin with the following 4 header bytes, which follow this
+ * scheme:
+ * |----------------------| |--| |----| |----|
+ * 31..               ..10  9,8   7..4   3..0
+ *        category         eflags #src   #dst
+ *
+ * 22 bits, category: it's a high level representation of the opcode of an instruction.
+ * Each bit represents one category following #dr_instr_category_t.
+ * Note that an instruction can belong to more than one category, hence multiple bits can
+ * be set.
+ * 2 bits, eflags: most significant bit set to 1 indicates the instruction reads at least
+ * one arithmetic flag, least significant bit set to 1 indicates the instruction writes
+ * at least one arithmetic flag.
+ * 4 bits, #src: number of source operands (read) that are registers. Registers used in
+ * memory reference operands of the instruction we are encoding (regardless on whether
+ * they are source or destination operands) are considered as source operands in the
+ * encoded instruction.
+ * 4 bits, #dst: number of destination operands (written) that are registers.
+ *
+ * We assume these encoded values to be little-endian.
+ * Note that we are only interested in register dependencies, hence operands that are
+ * not registers, such as immediates or memory references, are not present.
+ *
+ * Following the 4 header bytes are the bytes for the operation size and for encoding
+ * register operands, if any are present.
+ * The first byte contains the operation size encoded as a OPSZ_ enum value.
+ * Following the operation size are the register operand IDs.
+ * Each register operand is 1 byte.
+ * The destination operands go first, followed by the source operands.
+ * An instruction can have up to 8 operands (sources + destinations).
+ * Note that, because of 4 byte alignment, instructions with 1 to 3 operands will have a
+ * size of 8 bytes (4 header bytes + 1 byte for operation size + 3 operand-related/padding
+ * bytes), instructions with 4 to 7 operands will have a size of 12 bytes, while
+ * instructions with 8 operands will have the maximum size of 16 bytes.
+ * Instructions with no operands have only the 4 header bytes (no operation size byte nor
+ * operand-related bytes).
+ * For example, an instruction with 4 operands (1 dst, 3 src) has 12 bytes and would be
```

Comment:
Same here: the all-flush-left with ragged-right is just hard to visually split up into pieces.

Reply:
Done. https://github.com/DynamoRIO/dynamorio/pull/6691#discussion_r1554658906